zkir/Dialect/Field/IR/FieldTypes.cpp

```diff
@@ -48,6 +48,85 @@ unsigned getIntOrPrimeFieldBitWidth(Type type) {
   return cast<IntegerType>(type).getWidth();
 }
 
+ParseResult parseColonFieldType(AsmParser &parser, Type &type) {
+  if (failed(parser.parseColonType(type)))
+    return failure();
+
+  if (isa<PrimeFieldType>(type)) {
+    return success();
+  } else if (isa<ExtensionFieldTypeInterface>(type)) {
+    return success();
+  }
+  return parser.emitError(parser.getCurrentLocation(),
+                          "expected prime field or extension field type");
+}
+
+ParseResult validateAttribute(AsmParser &parser, Type type, Attribute attr,
+                              std::string_view attrName) {
+  if (auto pfType = dyn_cast<field::PrimeFieldType>(type)) {
+    if (!isa<IntegerAttr>(attr)) {
+      return parser.emitError(parser.getCurrentLocation(),
+                              "expected integer attribute for " +
+                                  std::string(attrName));
+    }
+    return success();
+  }
+  if (!isa<DenseIntElementsAttr>(attr)) {
+    return parser.emitError(parser.getCurrentLocation(),
+                            "expected dense int elements attribute for " +
+                                std::string(attrName));
+  }
+  return success();
+}
+
+Attribute maybeToMontgomery(Type type, Attribute attr) {
+  if (auto pfType = dyn_cast<field::PrimeFieldType>(type)) {
+    if (auto intAttr = dyn_cast<IntegerAttr>(attr)) {
+      if (pfType.isMontgomery()) {
+        return mod_arith::getAttrAsMontgomeryForm(pfType.getModulus(), intAttr);
+      } else {
+        return intAttr;
+      }
+    }
+    return nullptr;
```

Comment:
```suggestion
    auto intAttr = cast<IntegerAttr>(attr));
    if (pfType.isMontgomery()) {
      return mod_arith::getAttrAsMontgomeryForm(pfType.getModulus(), intAttr);
    } else {
      return intAttr;
    }
```

How about this? I am afraid `nullptr` will give us hard to track error.